File: 498-diagonal-traverse/diagonal-traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        vector<int> result;
        int row=mat.size(),col=mat[0].size();
        for(int d=0;d<row+col-1;++d){
            int r,c;
            if(d%2==0){ //Going up
                r=min(d,row-1);
                c=d-r;
                while(r>=0 && c<col){
                    result.push_back(mat[r][c]);
                    --r; ++c;
                }
            }
            else{ //going down
                c=min(d,col-1);
                r=d-c;
                while(c>=0 && r<row){
                    result.push_back(mat[r][c]);
                    ++r; --c;
                }
            }
        }
        return result;
    }
};

File: 498-diagonal-traverse/diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
       # diagonal up mean -1, +1
       # diagonal down mean +1, -1
       # passing boundary 
            # if past col then keep col add 1 to row
            # if past row then keep row add 1 to col

        sol = []
        m,n=len(mat),len(mat[0])
        up = True
        i,j=0,0

        while i<m and j<n:
            if up:
                sol.append(mat[i][j])
                i-=1
                j+=1

                if j>=n:
                    j-=1
                    i+=2
                    up=False
                elif i<0:
                    i+=1
                    up=False
            else:
                sol.append(mat[i][j])
                i+=1
                j-=1

                if i>=m:
                    i-=1
                    j+=2
                    up=True
                elif j<0:
                    j+=1
                    up=True

        return sol